testlapack/dlaqp2.go

```diff
@@ -0,0 +1,132 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"sort"
+	"testing"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+type Dlaqp2er interface {
+	Dlaqp2(m, n, offset int, a []float64, lda int, jpvt []int, tau, vn1, vn2, work []float64)
+}
+
+func Dlaqp2Test(t *testing.T, impl Dlaqp2er) {
+	rnd := rand.New(rand.NewSource(1))
+	for ti, test := range []struct {
+		m, n, offset int
+	}{
+		{m: 4, n: 3, offset: 0},
+		// {m: 4, n: 3, offset: 2},
+		// {m: 4, n: 3, offset: 4},
+		{m: 10, n: 10, offset: 0},
+		// {m: 10, n: 10, offset: 5},
+		// {m: 10, n: 10, offset: 10},
+	} {
+		m := test.m
+		n := test.n
+		jpiv := make([]int, n)
+
+		const (
+			all = iota
+			some
+			none
+		)
+		for _, free := range []int{all, some, none} {
+			for _, extra := range []int{0, 11} {
+				a := zeros(m, n, n+extra)
+				c := 1
+				for i := 0; i < m; i++ {
+					for j := 0; j < n; j++ {
+						a.Data[i*a.Stride+j] = float64(c)
+						c++
+					}
+				}
+				aCopy := cloneGeneral(a)
+				for j := range jpiv {
+					switch free {
+					case all:
+						jpiv[j] = -1
+					case some:
+						jpiv[j] = rnd.Intn(2) - 1
+					case none:
+						jpiv[j] = j
+					default:
+						panic("bad freedom")
+					}
+				}
+
+				tau := make([]float64, n)
+				vn1 := make([]float64, n)
```

Comment:
`lsticc` is assigned values from `vn2`. If these values are at least `len(vn2)` we get a bounds panic when we do `itemp := int(vn2[lsticc] + 0.5)`. So the partial norms are constrained to be less than `float64(len(vn2)) - 0.5`.

You can see this when you instrument so:

```diff
diff --git a/native/dlaqps.go b/native/dlaqps.go
index ce8cdb9..647e5e1 100644
--- a/native/dlaqps.go
+++ b/native/dlaqps.go
@@ -5,6 +5,7 @@
 package native
 
 import (
+       "fmt"
        "math"
 
        "github.com/gonum/blas"
@@ -201,7 +202,9 @@ func (impl Implementation) Dlaqps(m, n, offset, nb int, a []float64, lda int, jp
        //
        // Recomputation of difficult columns.
        for lsticc >= 0 {
+               fmt.Println(vn2, lsticc)
                itemp := int(vn2[lsticc] + 0.5)
+               fmt.Println(itemp)
 
                // NOTE: The computation of vn1[lsticc] relies on the fact that
                // Dnrm2 does not fail on vectors with norm below the value of
```
and set `vn2` to `columnNorms(m, n, a.Data, a.Stride)` in the tests as I've done for dlaqp2.go
```
$ go test -run Dlaqps
[12.884098726725126 14.628738838327795 -1] 2
0
[12.884098726725126 14.628738838327795 0.6175483551795395] 0
13
[0.5670198402783161 14.628738838327795 0.6175483551795395] 13
--- FAIL: TestDlaqps (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range
```